### DISCUSSION TITLE
Support for PostgreSQL check constraints

### DISCUSSION THREAD

**Discussion by jpholtmeyer** [kristiandupont/kanel#544] [General]
Would it be possible to support creating types that enforce PostgreSQL check constraints?
I'd like an opinion to see if it's worth investing effort into vs finding another solution.

As an example, lets say I define the column in a SQL table:
```
compounding_period VARCHAR CHECK (compounding_period IN (
        'DAILY', 'WEEKLY'))
```

I would like the ability to generate a type with a property that looks something like this in Typescript:
`type frequency = 'DAILY' | 'WEEKLY';`

And also Zod types:
z.enum(['DAILY', 'WEEKLY'])

Any help is greatly appreciated.

**Comment by kristiandupont** [comment]
Out of curiosity: is there any advantage to doing this over regular [Postgres enums](https://www.postgresql.org/docs/current/datatype-enum.html)? 

A more generic version that works with arbitrary constraints would have to parse those and create the equivalent Zod code. 
The schema does have information about constraints, so you should be able to do this. You could then create types similar to what I do in [Narrow Types](https://github.com/kristiandupont/narrow-types) and use those as well. It's an interesting topic to explore for sure. I think it should still be optional because there will be people who will not be interested in this.

**Reply by jpholtmeyer** [reply]
I initially had concerns about being able to modify Postgres enums adhoc, but I don't think that concern was warranted.
I modified my Postgres database to use enums rather than checks and then re-generated the types and I can see that the types were generated with enum-like constraints.
This looks like the behavior I was expecting.

I wouldn't want to rewrite any of the core code, but was thinking about creating my own pre-render hook and using it like a plugin.

Do you think that would be possible?
Or is there a better approach?

**Reply by kristiandupont** [reply]
It's definitely possible as a pre-render hook. The hard part with those is making them work with a variety of configurations (which is why I don't supply many of them out-of-the-box), but if you just need it to work for your particular config, it's typically quite easy. 

You will want to inspect the [checks](https://kristiandupont.github.io/extract-pg-schema/api/extract-pg-schema.tabledetails.checks.html) part of the `table` and then modify the output accordingly. I hope the [documentation](https://kristiandupont.github.io/kanel/preRenderHooks.html) is helpful but feel free to ask here as well :-)

**Reply by jpholtmeyer** [reply]
This is really good info, and the documentation has been helpful.
Just wanted to make sure I was thinking in the right direction.

Thanks a bunch!
